ethcore/src/blockchain/blockchain.rs

```diff
@@ -979,6 +995,34 @@ impl BlockChain {
 		}
 	}
 
+	/// Mark a block to be considered finalized. Panic if the hash does not exist.
+	pub fn mark_finalized(&self, batch: &mut DBTransaction, block_hash: H256) {
+		let mut block_details = self.block_details(&block_hash).unwrap_or_else(|| panic!("Invalid block hash: {:?}", block_hash));
+		block_details.finalized = true;
+
+		self.prepare_details_update(batch, block_hash, block_details);
+	}
+
+	/// Update metadata detail for an existing block.
+	pub fn update_metadatas(&self, batch: &mut DBTransaction, block_hash: H256, metadatas: HashMap<Bytes, Bytes>) {
```

Comment:
Wouldn't it be better to have metadata in some more structured way? Seems that `HashMap<Bytes, Bytes>` is even worse than `HashMap<String, String>` (the latter is at least human readable)

Reply:
Maybe `HashMap<String, Bytes>`? For many cases (for example, to store Casper score info `U256`), the value itself is readable anyway.

Reply:
also: `metadata` is already plural

Reply:
What I mean is maybe to have `T: TryFrom<HashMap<Bytes, Bytes>>` or sth (if we want to be super generic), so that the caller can decide on the expected type of metadata, otherwise we might be passing a lot of weakly-typed structures in the codebase and imho would be best to avoid that.

Alternatively maybe an `enum` with possible types of stored metadata if we are ok with having less generic solution and a single place where all kinds of metadata is defined.